rados/read_op.go

```diff
@@ -0,0 +1,84 @@
+package rados
+
+// #cgo LDFLAGS: -lrados
+// #include <errno.h>
+// #include <stdlib.h>
+// #include <rados/librados.h>
+//
+import "C"
+
+import (
+	"unsafe"
+)
+
+// ReadOp manages a set of discrete object read actions that will be performed
+// together atomically.
+type ReadOp struct {
+	operation
+	op C.rados_read_op_t
+}
+
+// CreateReadOp returns a newly constructed read operation.
+func CreateReadOp() *ReadOp {
+	return &ReadOp{
+		op: C.rados_create_read_op(),
+	}
+}
+
+// Release the resources associated with this read operation.
+func (r *ReadOp) Release() {
+	C.rados_release_read_op(r.op)
+	r.op = nil
+	r.free()
+}
+
+// Operate will perform the operation(s).
+func (r *ReadOp) Operate(ioctx *IOContext, oid string, flags OperationFlags) error {
+	if err := ioctx.validate(); err != nil {
```

Comment:
It can, but there's only so much defensive program even I'm willing to do. I'm perfectly happy with this panicking if the caller passes in nil as ioctx. The validate is there to make sure our internal state (that the caller is not allowed to touch is valid).

Reply:
Are you OK with this now?

Reply:
I'd like to see the check for a non-nil ioctx too, of course :wink:

Reply:
I think I'd like to table this part of the PR discussion and make it a topic for the next time we talk. I think you'll all agree that we want to make calls in go-ceph consistent and so this part of things has wider reaching implications (IMO). If you can both stomach not checking it for now we can discuss the more "meta" aspects of this later and decide what approach we want throughout go-ceph.

Reply:
FWIW I have added a test to ensure passing a nil ioctx panics.